scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+  assert(ncol(curr_X) == 1)
+  assert(ncol(X) == 1)
+
+  left = matrix(0, rows=0, cols=1)
+  right = matrix(0, rows=0, cols=1)
+
+  for(i in 1:nrow(curr_X))
+  {
+    if (as.scalar(curr_X[i,]) == 1) # categorical true
+    {
+      left = rbind(left, curr_X[i,])
+      right = rbind(right, matrix(0, rows=1, cols=1))
+    }
+    else if(as.scalar(curr_X[i,]) == 0)  # categorical false
+    {
+      left = rbind(left, curr_X[i,])
+      right = rbind(right, matrix(1, rows=1, cols=1))
+    }
+    else  # replace with NaN if not used for the current samples
+    {
+      assert(as.scalar(curr_X[i,]) == 'NaN')
+      left = rbind(left, curr_X[i,])
+      right = rbind(right, curr_X[i,])
+    }
+  }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    row_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateSimilarityScore = function (matrix[Double] row_vector, Double lambda) return (Double similarity_score) {
+  similarity_score = (sum(row_vector)^2) / (nrow(row_vector) + lambda);
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    row_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateSimilarityScoreClassification = function (matrix[Double] row_vector, matrix[Double] predictions, Double lambda) return (Double similarity_score) {
+  nominator = (sum(row_vector)^2)
+  d =  predictions * (1 - predictions)
+  denominator = sum(d) + lambda
+  similarity_score = nominator / denominator;
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    residuals_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateOutputValue = function (matrix[Double] residuals_vector, Double lambda) return (Double output_value) {
+  output_value = (sum(residuals_vector)) / (nrow(residuals_vector) + lambda);
+  if(output_value == 'NaN') # just in case we have a node with no sample inside
+    output_value = 0.0
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    residuals_vector: a 1xn matrix (one feature with all residuals)
+# OUTPUT:   similarity_score: the similarity score of the residuals
+calculateOutputValueClassification = function (matrix[Double] residuals_vector, matrix[Double] predictions, Double lambda) return (Double output_value) {
+    nominator = (sum(residuals_vector))
+    d =  predictions * (1 - predictions)
+    denominator = sum(d) + lambda
+    if(denominator == 0)
+      output_value = 0
```

Comment:
Are you sure you want to assign it to 0 not Infinity?

Reply:
yes - should be fine